src/Simulation/Simulators/ToffoliSimulator/Processor.cs

```diff
@@ -0,0 +1,429 @@
+                    }
+                    SetValue(qubit, false);
+                }
+            }
+        }
+
+        // The borrowed qubits' states are stored on a stack together with their current value.
+        // The value is used for comparison when the qubit is returned.
+        public override void OnBorrowQubits(IQArray<Qubit> qubits, long allocatedForBorrowingCount) {
+            foreach (var qubit in qubits) {
+                Stack<bool> stack;
+                if (!borrowedQubitStates.TryGetValue(qubit, out stack)) {
+                    stack = new Stack<bool>();
+                    borrowedQubitStates.Add(qubit, stack);
+                }
+                stack.Push(GetValue(qubit));
+            }
+        }
+
+        // A qubit must be returned in its original value when borrowed.
+        public override void OnReturnQubits(IQArray<Qubit> qubits, long releasedOnReturnCount) {
+            foreach (var qubit in qubits) {
+                var stack = borrowedQubitStates[qubit];
+                if (stack.Pop() != GetValue(qubit)) {
+                    throw new ExecutionFailException("Borrowed qubit returned not in the original state");
+                }
+
+                if (stack.Count == 0) {
+                    borrowedQubitStates.Remove(qubit);
+                }
+            }
+        }
+
+        // An X operation inverts the bit in the `simulationValues` variable at
+        // the position of the qubit's index.
+        public override void X(Qubit qubit) {
+            InvertValue(qubit);
+        }
+
+        // If the simulation values of all control qubits are assigned true,
+        // the simulation value of the target qubit is inverted.
+        public override void ControlledX(IQArray<Qubit> controls, Qubit qubit) {
+            if (controls.All(control => GetValue(control))) {
+                InvertValue(qubit);
+            }
+        }
+
+        // SWAP interchanges the simulation values at two locations.
+        public override void SWAP(Qubit qubit1, Qubit qubit2)
+        {
+            var tmp = GetValue(qubit1);
+            SetValue(qubit1, GetValue(qubit2));
+            SetValue(qubit2, tmp);
+        }
+
+        /// <summary>
+        /// Make sure that a rotation is either the identity or an X, up to global phase.
+        /// Assumes that the rotation is exp(i*theta*sigma(axis)), where sigma(axis) is
+        /// the appropriate Pauli matrix normalized to +/-1 eigenvalues.
+        /// Throws an exception if the rotation is invalid.
+        /// </summary>
+        /// <param name="axis">The rotation axis</param>
+        /// <param name="theta">The rotation angle, properly normalized</param>
+        /// <returns>A pair of flags. The first indicates whether the rotation is
+        /// equivalent to X; the second whether or not the rotation is exactly equal
+        /// to the identity (no phase), and thus may be controlled.</returns>
```

Comment:
Why is it valuable to support in Toffoli simulator rotations that are equivalent to X or I? Do we expect the users to write their code for running on a Toffoli simulator in terms of such rotations?

Reply:
I don't know, and doubt that this is a common scenario. This is supported by the current implementation and I wanted to make sure to take over all currently supported features.

Reply:
Supporting this case does complicate the code and sends a somewhat confusing message to the users: "Rotation gates are supported _sometimes_", so I was wondering whether we have any data on this functionality actually being valuable...

---
In reply to: [444429979](https://github.com/microsoft/qsharp-runtime/pull/276#discussion_r444429979) [](ancestors = 444429979)

Reply:
I wouldn't want to remove any functionality in this conversion PR. Maybe this can be another discussion and another PR. I think there are sufficiently many warnings to report to the user which kind of rotations are not supported.

Reply:
I agree with Mathias that backwards compatibility is important here, regardless of whether the functionality us more or less useful.

---
In reply to: [444679013](https://github.com/microsoft/qsharp-runtime/pull/276#discussion_r444679013) [](ancestors = 444679013)